### PR TITLE
Toast animations

### PR DESCRIPTION
**Pull request by fbasso**
Inspired by #2817, not cherry-picked because the toast animations are finally quite different compared to the alert.

This PR has 3 commits:
- toast animations implementation
- fix some test, due to a change on ngbTransition (the startFn function and its optional callback are now run even if animation = false)
- The changes in the demo examples

### Toast animations

The toast animations are defined as follow:

- The `fade` class exists only when the animations are defined,
- When visible, the toast contains the class `show`,
- When not visible, the toast contains the class `hide` (if, for any reason, the developer choose to keep the toast in the dom),
- When closing, the `show` class is removed to start the transition, then the `hide` class is added once finished,
- When opening, the `hide` class is removed, the `showing` class is added. Then the `showing` class is replaced the the `show` class,
- `show` (new api) and `hide` return the transition observer (to be able to implement code after the animations without the events),
- New events are available: `shown` and `hidden` (`hide`is deprecated, to stick with the official Bootstrap api).

### CODE REVIEW

File: src/toast/toast.ts [hunk 1]
```diff
@@ -118,7 +144,19 @@ export class NgbToast implements AfterContentInit,
 
   hide() {
```
Comment (maxokorokov) [inline src/toast/toast.ts:145]: Could we add documentation for both `.hide()` and `.show()` methods? At the moment there is no documentation for this on the demo site?

**Comment by maxokorokov** [comment]
Also ideally the `BREAKING CHANGE: ` commit message should be for the `feat(toast): add animations` commit.

File: src/toast/toast.ts [hunk 2]
```diff
@@ -97,17 +115,25 @@ export class NgbToast implements AfterContentInit,
...
       this.ariaLive = config.ariaLive;
     }
     this.delay = config.delay;
     this.autohide = config.autohide;
+    this.animation = config.animation;
   }
 
-  ngAfterContentInit() { this._init(); }
+  ngAfterContentInit() {
+    this._zone.onStable.asObservable().pipe(take(1)).subscribe(() => {
```
Comment (benouat) [inline src/toast/toast.ts:132]: Why do we need zone here ?

File: src/util/transition/ngbFadingTransition.ts [hunk 3]
```diff
@@ -3,3 +3,21 @@ import {NgbTransitionStartFn} from './ngbTransition';
...
+    classList.remove('showing');
+    classList.add('show');
+  };
+};
+
+export const ngbToastFadeOutTransition: NgbTransitionStartFn<NgbToastCtx> = ({classList}: HTMLElement) => {
+  classList.remove('show');
+
+  return () => { classList.add('hide'); };
+};
```
Comment (benouat) [inline src/util/transition/ngbFadingTransition.ts:23]: After 3 widgets conversion, I am still not convinced about the need for this abstraction.
These functions should be inlined in each component, even if used several times.

File: src/util/transition/ngbFadingTransition.ts [hunk 4]
```diff
@@ -3,3 +3,21 @@ import {NgbTransitionStartFn} from './ngbTransition';
...
+  classList.remove('hide');
+  classList.add('showing');
+
+  return () => {
+    classList.remove('showing');
+    classList.add('show');
+  };
+};
+
+export const ngbToastFadeOutTransition: NgbTransitionStartFn<NgbToastCtx> = ({classList}: HTMLElement) => {
```
Comment (benouat) [inline src/util/transition/ngbFadingTransition.ts:19]: Don't forget to remove the context in the types also

File: src/util/transition/ngbFadingTransition.ts [hunk 5]
```diff
@@ -3,3 +3,21 @@ import {NgbTransitionStartFn} from './ngbTransition';
 export const ngbAlertFadingTransition: NgbTransitionStartFn = ({classList}: HTMLElement) => {
   classList.remove('show');
 };
+
+export interface NgbToastCtx { direction: 'show' | 'hide'; }
+
+export const ngbToastFadeInTransition: NgbTransitionStartFn<NgbToastCtx> = ({classList}: HTMLElement) => {
```
Comment (benouat) [inline src/util/transition/ngbFadingTransition.ts:9]: Don't forget to remove the context in types also

File: src/util/transition/ngbTransition.ts [hunk 6]
```diff
@@ -28,6 +28,7 @@ export const ngbRunTransition =
 
           // If animations are disabled, we have to emit a value and complete the observable
           if (!options.animation) {
+            (startFn(element, null !) || noopFn)();
```
Comment (benouat) [inline src/util/transition/ngbTransition.ts:31]: Why is this needed? it's not clear to me. Is it because even without animations we need to remove the `hidden` class? If yes, let's not put it in case animation is false.

File: src/toast/toast.spec.ts [hunk 7]
```diff
@@ -1,21 +1,21 @@
...
-const getToastElement = (fixture: ComponentFixture<TestComponent>): Element =>
-    getElementWithSelector(fixture, 'ngb-toast');
-const getToastHeaderElement = (fixture: ComponentFixture<TestComponent>): Element =>
-    getElementWithSelector(fixture, 'ngb-toast .toast-header');
-const getToastBodyElement = (fixture: ComponentFixture<TestComponent>): Element =>
-    getElementWithSelector(fixture, 'ngb-toast .toast-body');
+const getToastElement = (element: HTMLElement): Element => getElementWithSelector(element, 'ngb-toast');
+const getToastHeaderElement = (element: HTMLElement): Element =>
+    getElementWithSelector(element, 'ngb-toast .toast-header');
+const getToastBodyElement = (element: HTMLElement): Element => getElementWithSelector(element, 'ngb-toast .toast-body');
```
Comment (benouat) [inline src/toast/toast.spec.ts:18]: Why? Because of that change, you had to put `fixture.nativeElement` everywhere in the file...

File: src/toast/toast.ts (hunk 2 above)
Comment (maxokorokov) [reply src/toast/toast.ts:132]: I guess we need to add `.show` class when `.fade` is already in place, otherwise the browser won't triggger the fade in transition. And we need this to be synchronous.

File: src/util/transition/ngbTransition.ts (hunk 6 above)
Comment (maxokorokov) [reply src/util/transition/ngbTransition.ts:31]: I think that's the explanation → https://github.com/ng-bootstrap/ng-bootstrap/pull/3746#issuecomment-635186699

TLDR:
- `[class.fade]: 'animation'` always bindings
- `.show` always added/removed via the `ngbRunTransition`

Othwise you have to "duplicate" adding show class via binding and via `ngbRunTransition`

File: demo/src/app/components/toast/demos/closeable/toast-closeable.ts [hunk 8]
```diff
@@ -6,7 +6,8 @@ export class NgbdToastCloseable {
   show = true;
 
   close() {
+    console.log('close called');
```
[the remark this answers is not in the document]
Comment (fbasso) [reply demo/src/app/components/toast/demos/closeable/toast-closeable.ts:9]: Missed it ! 😕

File: src/toast/toast.ts (hunk 2 above)
Comment (fbasso) [reply src/toast/toast.ts:132]: Exactly. No animation is shown without it.

File: src/util/transition/ngbFadingTransition.ts (hunk 3 above)
Comment (fbasso) [reply src/util/transition/ngbFadingTransition.ts:23]: It's a matter of coding preferences. Even if simple, I prefer to abstract this specific action done in the widget. 

`ngbToastFadeOutTransition` is a kind of documentation. When I read `
this._element.nativeElement, ngbToastFadeOutTransition, {animation: this.animation, runningTransition: 'stop'});` in the toast, I can easily understand what the code intend to do (a fade out), and separate the fade out code form the rest of the method code. In the fade out code, I can also easily know what are the Bootstrap specs on this transition.

But it's up to the main project decision and guidelines. No worries, I can change it if needed ;)

File: src/util/transition/ngbFadingTransition.ts (hunk 4 above)
Comment (fbasso) [reply src/util/transition/ngbFadingTransition.ts:19]: Sure, thanks !

File: src/toast/toast.spec.ts (hunk 7 above)
Comment (fbasso) [reply src/toast/toast.spec.ts:18]: It's because I have this error when I use it in the animation part:

```
Argument of type 'ComponentFixture<TestAnimationComponent>' is not assignable to parameter of type 'ComponentFixture<TestComponent>'
```

The possible fixes I found were:

- Remove the type in the signature of getToastElement:
`getToastElement(fixture) ...`, but we lost the type feature in this function.
- Extract TestAnimationComponent from its `it` context and move it to the top, then define getToastElement as follow : `const getToastElement = (fixture: ComponentFixture<TestComponent | TestAnimationComponent>)`

I ended up by follow what is done in alert.specs.ts.

What do you prefer ? Any other idea ?

File: src/util/transition/ngbFadingTransition.ts (hunk 3 above)
Comment (maxokorokov) [reply src/util/transition/ngbFadingTransition.ts]: Added this to the list → https://github.com/ng-bootstrap/ng-bootstrap/issues/3747#issue-626345524

**Comment by maxokorokov** [comment]
LGTM, will merge when CI is green!

@fbasso, I just also pushed this (https://github.com/ng-bootstrap/ng-bootstrap/pull/3749/commits/91c5356e16467e4034a589af78487c75e696feb8), because we forgot to fix documentation for the `(shown)` and `(hidden)`.
Also `.hide()` and `.show()` return types were missing in generated docs. You have to put them explicitly, so they would appear:

Screen Shot 2020-06-03 at 17 38 53
